diff -U 3 -r binutils-2.38-original/ld/emulparams/elf32lriscv-defs.sh binutils-2.38/ld/emulparams/elf32lriscv-defs.sh
--- binutils-2.38-original/ld/emulparams/elf32lriscv-defs.sh	2022-01-22 07:14:09.000000000 -0500
+++ binutils-2.38/ld/emulparams/elf32lriscv-defs.sh	2022-06-28 22:05:05.458964533 -0400
@@ -27,7 +27,7 @@
 esac
 
 IREL_IN_PLT=
-TEXT_START_ADDR=0x10000
+TEXT_START_ADDR=0x1000000
 MAXPAGESIZE="CONSTANT (MAXPAGESIZE)"
 COMMONPAGESIZE="CONSTANT (COMMONPAGESIZE)"
 
diff -U 3 -r binutils-2.38-original/ld/scripttempl/elf.sc binutils-2.38/ld/scripttempl/elf.sc
--- binutils-2.38-original/ld/scripttempl/elf.sc	2022-01-22 07:14:09.000000000 -0500
+++ binutils-2.38/ld/scripttempl/elf.sc	2022-06-28 22:05:05.462964444 -0400
@@ -613,6 +613,7 @@
 
   ${RELOCATING+/* Adjust the address for the data segment.  We want to adjust up to
      the same address within the page on the next page up.  */}
+  ${RELOCATING+. = (. + 0xFFFF) & ~0xFFFF;}
   ${CREATE_SHLIB-${CREATE_PIE-${RELOCATING+. = ${DATA_ADDR-${DATA_SEGMENT_ALIGN}};}}}
   ${CREATE_SHLIB+. = ${SHLIB_DATA_ADDR-${DATA_SEGMENT_ALIGN}};}
   ${CREATE_PIE+. = ${SHLIB_DATA_ADDR-${DATA_SEGMENT_ALIGN}};}
